src/main/java/io/asyncer/r2dbc/mysql/message/server/OkMessage.java

```diff
@@ -164,26 +182,79 @@ static OkMessage decode(boolean isEndOfRows, ByteBuf buf, ConnectionContext cont
+            Map<String, String> systemVariables = Collections.emptyMap();
+
+            while (VarIntUtils.checkNextVarInt(buf) >= 0) {
+                int stateInfoSize = (int) VarIntUtils.readVarInt(buf);
+                ByteBuf stateInfo = buf.readSlice(stateInfoSize);
+
+                while (stateInfo.isReadable()) {
+                    if (stateInfo.readByte() == SESSION_TRACK_SYSTEM_VARIABLES) {
+                        systemVariables = readServerVariables(stateInfo, context);
+                    } else {
+                        // Ignore other state info
+                        int skipBytes = (int) VarIntUtils.readVarInt(stateInfo);
 
-            if (size > sizeAfterVarInt) {
-                information = buf.toString(readerIndex, buf.writerIndex() - readerIndex, charset);
-            } else {
-                // JVM does NOT support strings longer than Integer.MAX_VALUE
-                information = buf.toString(buf.readerIndex(), (int) size, charset);
+                        stateInfo.skipBytes(skipBytes);
+                    }
+                }
             }
 
             // Ignore session track, it is not human-readable and useless for R2DBC client.
             return new OkMessage(isEndOfRows, affectedRows, lastInsertId, serverStatuses, warnings,
-                information);
+                information, systemVariables);
         }
 
         // Maybe have no human-readable message
-        return new OkMessage(isEndOfRows, affectedRows, lastInsertId, serverStatuses, warnings, "");
+        return new OkMessage(isEndOfRows, affectedRows, lastInsertId, serverStatuses, warnings, "",
+            Collections.emptyMap());
+    }
+
+    private static Map<String, String> readServerVariables(ByteBuf buf, ConnectionContext context) {
+        // All lengths should NOT be greater than Integer.MAX_VALUE
+        Map<String, String> map = new HashMap<>();
+        Charset charset = context.getClientCollation().getCharset();
+        int size = (int) VarIntUtils.readVarInt(buf);
+        ByteBuf sessionVar = buf.readSlice(size);
+
+        while (sessionVar.readableBytes() > 0) {
+            int variableSize = (int) VarIntUtils.readVarInt(sessionVar);
+            String variable = sessionVar.toString(sessionVar.readerIndex(), variableSize, charset);
+
+            sessionVar.skipBytes(variableSize);
+
+            int valueSize = (int) VarIntUtils.readVarInt(sessionVar);
+            String value = sessionVar.toString(sessionVar.readerIndex(), valueSize, charset);
+
+            sessionVar.skipBytes(valueSize);
+            map.put(variable, value);
+        }
+
+        switch (map.size()) {
+            case 0:
+                return Collections.emptyMap();
+            case 1: {
+                Map.Entry<String, String> entry = map.entrySet().iterator().next();
+                return Collections.singletonMap(entry.getKey(), entry.getValue());
+            }
+            default:
+                return map;
```

Comment:
Could we consider returning `Collections.unmodifiableMap(map)` to ensure the map remains read-only?

Reply:
I consider that this map should only have the `get` method called.

Maybe we can change `Map<String, String> getSystemVariables()` to `@Nullable String getSystemVariable(String key)`, then this map will never be changed outer than `OkMessage`

Reply:
make sense. lgtm